listener.go

```diff
@@ -32,16 +36,22 @@ func newLocalhostListener(ports []int) (*localhostListener, error) {
 	return nil, xerrors.Errorf("no available port (%s)", strings.Join(errs, ", "))
 }
 
-func newLocalhostListenerAt(port int) (*localhostListener, error) {
-	l, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
+// newLocalhostListenerAt starts a TCP listener on localhost and given port.
+//
+// If an address is given, it will bind the address. It defaults to localhost.
+// Note that this always returns a localhost URL regardless of the address.
+func newLocalhostListenerAt(address string, port int) (*localhostListener, error) {
+	if address == "" {
+		address = "localhost"
+	}
+	l, err := net.Listen("tcp", fmt.Sprintf("%s:%d", address, port))
 	if err != nil {
 		return nil, xerrors.Errorf("could not listen: %w", err)
 	}
-	addr := l.Addr().String()
-	_, p, err := net.SplitHostPort(addr)
-	if err != nil {
-		return nil, xerrors.Errorf("could not parse the address %s: %w", addr, err)
+	addr, ok := l.Addr().(*net.TCPAddr)
```

Comment:
I'm not sure what that does. What is the difference between this two blocks?

```
addr, ok := l.Addr().(*net.TCPAddr)
```
and
```
addr := l.Addr()
```

Reply:
We need to cast it to net.TCPAddr to extract the port number of the local server. We can still split host and port of the address but type casting is more simple.